pkg/operator/csr/csr_approver.go

```diff
@@ -0,0 +1,275 @@
+	return &ServiceAccountApprover{
+		saName:          serviceaccount.MakeUsername(saNamespace, saName),
+		saGroups:        sets.NewString(append(serviceaccount.MakeGroupNames(saNamespace), "system:authenticated")...),
+		expectedSubject: expectedSubject,
+	}
+}
+
+func (a *ServiceAccountApprover) Approve(csrObj *certapiv1.CertificateSigningRequest, x509CSR *x509.CertificateRequest) (approvalStatus CSRApprovalDecision, denyReason string, err error) {
+	if csrObj == nil || x509CSR == nil {
+		return CSRDenied, "Error", fmt.Errorf("received a 'nil' CSR")
+	}
+
+	if csrObj.Spec.Username != a.saName {
+		return CSRDenied, fmt.Sprintf("CSR %q was created by an unexpected user: %q", csrObj.Name, csrObj.Spec.Username), nil
+	}
+
+	if csrGroups := sets.NewString(csrObj.Spec.Groups...); !csrGroups.IsSuperset(a.saGroups) {
+		return CSRDenied, fmt.Sprintf("CSR %q was created by a user with unexpected groups: %v", csrObj.Name, csrGroups.List()), nil
+	}
+
+	if expectedSubject := a.expectedSubject; x509CSR.Subject.String() != expectedSubject {
+		return CSRDenied, fmt.Sprintf("expected the CSR's subject to be %q, but it is %q", expectedSubject, x509CSR.Subject.String()), nil
+	}
+
+	return CSRApproved, "", nil
+
+}
+
+type CSRFilter interface {
+	Matches(csr *certapiv1.CertificateSigningRequest) bool
+}
+
+type AndFilter struct {
+	a, b CSRFilter
+}
+
+func NewAndFilter(a, b CSRFilter) *AndFilter {
+	return &AndFilter{a, b}
+}
+
+func (f *AndFilter) Matches(csr *certapiv1.CertificateSigningRequest) bool {
+	return f.a.Matches(csr) && f.b.Matches(csr)
+}
+
+type OrFilter struct {
+	a, b CSRFilter
+}
+
+func NewOrFilter(a, b CSRFilter) *OrFilter {
+	return &OrFilter{a, b}
+}
+
+func (f *OrFilter) Matches(csr *certapiv1.CertificateSigningRequest) bool {
+	return f.a.Matches(csr) || f.b.Matches(csr)
+}
+
+type LabelFilter struct {
+	labelSelector labels.Selector
+}
+
+func NewLabelFilter(selector labels.Selector) *LabelFilter {
+	return &LabelFilter{selector}
+}
+
+func (f *LabelFilter) Matches(csr *certapiv1.CertificateSigningRequest) bool {
+	return f.labelSelector.Matches(labels.Set(csr.Labels))
+}
+
+type NamesFilter struct {
+	names sets.String
+}
+
+func NewNamesFilter(names ...string) *NamesFilter {
+	return &NamesFilter{sets.NewString(names...)}
+}
+
+func (f *NamesFilter) Matches(csr *certapiv1.CertificateSigningRequest) bool {
+	return f.names.Has(csr.Name)
+}
+
+type RequestCommonNameFilter struct {
+	commonNames sets.String
+}
+
+func NewRequestCommonNameFilter(commonNames ...string) *RequestCommonNameFilter {
+	return &RequestCommonNameFilter{sets.NewString(commonNames...)}
+}
+
+func (f *RequestCommonNameFilter) Match(csr *certapiv1.CertificateSigningRequest) bool {
+	x509CSR, err := x509.ParseCertificateRequest(csr.Spec.Request)
+	if err != nil {
+		klog.V(4).Infof("failed to parse the CSR .spec.request of %q: %v", csr.Name, err)
+		return false
+	}
+
+	return f.commonNames.Has(x509CSR.Subject.CommonName)
```

Comment:
difference between subject and and common name filters?  What else does the subject include?

Reply:
There may not be too many uses inside kube for other subject fields, but there's a few other possible fields, typically Organization (o), OrganizationalUnit (ou), Location (l).

Alternatively, we can just match against the string representation of the full subject.